src/Host/Broker/Impl/Startup/CommonStartup.cs

```diff
@@ -142,16 +156,37 @@ internal class CommonStartup {
             return webHost;
         }
 
+        private static void ServiceExit() {
+            using (var serviceController = new ServiceController(Resources.Text_ServiceName)) {
+                serviceController.Stop();
+            }
+        }
+
         internal static void Exit() {
             _cts.Cancel();
 
             Task.Run(async () => {
-                // Give cooperative cancellation 10 seconds to shut the process down gracefully,
-                // but if it didn't work, just terminate it.
-                await Task.Delay(10000);
-                _logger.LogCritical(Resources.Critical_TimeOutShutdown);
-                Environment.Exit((int)BrokerExitCodes.Timeout);
+                if (IsService) {
+                    ServiceExit();
+                } else {
+                    // Give cooperative cancellation 10 seconds to shut the process down gracefully,
+                    // but if it didn't work, just terminate it.
+                    await Task.Delay(10000);
```

Comment:
This seem just to wait for 10 sec and then log critical. Is it expected to check for successful termination somewhere?